Tribler/Core/TorrentChecker/session.py

```diff
@@ -198,309 +185,384 @@ def __init__(self, tracker_url, tracker_address, announce_page, on_result_callba
+                                                  {"info_hash" : self._infohash_list})
 
-        self._logger.debug(u"%s HTTP SCRAPE message sent", self)
+        agent = RedirectAgent(Agent(reactor, connectTimeout=15.0))
+        self.request = agent.request('GET', bytes(url))
+        self.request.addCallback(self.cbresponse)
+        self.request.addErrback(self.on_error)
+
+        self._logger.debug(u"%s HTTP SCRAPE message sent: %s", self, url)
 
         # no more requests can be appended to this session
         self._action = TRACKER_ACTION_SCRAPE
         self._is_initiated = True
+        self._last_contact = int(time.time())
 
-    def _handle_response(self):
-        try:
-            # TODO: this buffer size may be changed
-            response = self._socket.recv(8192)
-        except Exception as e:
-            self._logger.info(u"%s Failed to receive HTTP SCRAPE response: %s", self, e)
-            self._is_failed = True
-            return
+        # Return deferred that will evaluate when the whole chain is done.
+        self.result_deferred = Deferred()
+        return self.result_deferred
 
-        self._logger.debug(u"%s Got response", self)
+    def on_error(self, failure):
+        self._logger.info("Error when querying http tracker: %s %s" % (str(failure), self.tracker_url))
+        self.failed()
 
-        if not response:
-            self._is_failed = True
+    def cbresponse(self, response):
+        # Check if this one was OK.
+        if response.code != 200:
+            # error response code
+            self._logger.warning(u"%s HTTP SCRAPE error response code [%s, %s]", self, response.code, response.phrase)
+            self.failed()
             return
 
-        # for the header message, we need to parse the content length in case
-        # if the HTTP packets are partial.
-        if not self._message_buffer:
-            # append the header part
-            if not self._header_buffer:
-                self._header_buffer = response
-            else:
-                self._header_buffer += response
-
-            # check if the header part is over
-            if self._header_buffer.find('\r\n\r\n') != -1:
-                self._header_buffer, self._message_buffer = self._header_buffer.split('\r\n\r\n', 1)
+        # All ok, parse the body
+        d = readBody(response)
+        d.addCallbacks(self._process_scrape_response, self.on_error)
 
-                self._received_length = len(self._message_buffer)
-                self._process_header()
+    def failed(self):
+        """
+        This method handles everything that needs to be done when one step
+        in the session has failed and thus no data can be obtained.
+        """
+        self._is_failed = True
+        if self.result_deferred:
+            self.result_deferred.errback(ValueError("HTTP tracker has failed for url %s" % self._tracker_url))
+        self.cleanup()
```

Comment:
You are ignoring the deferred the cleanup returns. Either return it so it joins the callback chain or remove it. Otherwise it will get really confusing in the future if cleanup really does async stuff.